plugins/modules/vault_kv2_write.py

```diff
@@ -0,0 +1,229 @@
+    type: str
+    default: secret
+  path:
+    type: str
+    required: true
+    description:
+      - Vault KVv2 path to be written to.
+      - This is relative to the I(engine_mount_point), so the mount path should not be included.
+  data:
+    type: dict
+    required: true
+    description:
+      - KVv2 secret data to write/patch.
+  cas:
+    type: int
+    description:
+      - Perform a check-and-set operation.
+  read_before_write:
+    type: bool
+    default: false
+    description:
+      - Read the secret first and write only when I(data) differs from the read data.
+      - Requires C(read) permission on the secret if C(true).
+      - If C(false), this module will always write to I(path).
+"""
+
+EXAMPLES = r"""
+- name: Write/create a secret
+  community.hashi_vault.vault_kv2_write:
+    url: https://vault:8200
+    path: hello
+    data:
+      foo: bar
+
+- name: Create a secret with CAS (the secret must not exist)
+  community.hashi_vault.vault_kv2_write:
+    url: https://vault:8200
+    path: caspath
+    cas: 0
+    data:
+      foo: bar
+
+- name: Update a secret with CAS
+  community.hashi_vault.vault_kv2_write:
+    url: https://vault:8200
+    path: caspath
+    cas: 2
+    data:
+      hello: world
+"""
+
+RETURN = r"""
+raw:
+  type: dict
+  description: The raw Vault response.
+  returned: changed
+  sample:
+    auth:
+    data:
+      created_time: "2023-02-21T19:51:50.801757862Z"
+      custom_metadata:
+      deletion_time: ""
+      destroyed: false
+      version: 1
+    lease_duration: 0
+    lease_id: ""
+    renewable: false
+    request_id: 52eb1aa7-5a38-9a02-9246-efc5bf9581ec
+    warnings: null
+    wrap_info: null
+"""
+
+import traceback
+
+from ansible.module_utils._text import to_native
+from ansible.module_utils.basic import missing_required_lib
+from ..module_utils._hashi_vault_common import (
+    HashiVaultValueError,
+)
+from ..module_utils._hashi_vault_module import (
+    HashiVaultModule,
+)
+
+try:
+    import hvac
+except ImportError:
+    HAS_HVAC = False
+    HVAC_IMPORT_ERROR = traceback.format_exc()
+else:
+    HVAC_IMPORT_ERROR = None
+    HAS_HVAC = True
+
+
+def run_module():
+    argspec = HashiVaultModule.generate_argspec(
+        engine_mount_point=dict(type="str", default="secret"),
+        path=dict(type="str", required=True),
+        data=dict(type="dict", required=True, no_log=True),
+        cas=dict(type="int"),
+        read_before_write=dict(type="bool", default=False),
+    )
+
+    module = HashiVaultModule(
+        argument_spec=argspec,
+        supports_check_mode=True,
+    )
+
+    if not HAS_HVAC:
+        module.fail_json(
+            msg=missing_required_lib("hvac"),
+            exception=HVAC_IMPORT_ERROR,
+        )
+
+    mount_point = module.params.get("engine_mount_point")
+    path = module.params.get("path")
+    cas = module.params.get("cas")
+    data = module.params.get("data")
+    read = module.params.get("read_before_write")
```

Comment:
Let's change this `read` variable (and the other references to it) to also be `read_before_write`